cluster/images/conformance/run_e2e.sh

```diff
@@ -39,23 +39,32 @@ saveResults() {
 # We get the TERM from kubernetes and handle it gracefully
 trap shutdown TERM
 
-ginkgo_args=(
-    "--focus=${E2E_FOCUS}"
-    "--skip=${E2E_SKIP}"
-    "--noColor=true"
-)
-
+ginkgo_args=()
 if [[ -n ${E2E_DRYRUN:-} ]]; then
     ginkgo_args+=("--dryRun=true")
 fi
 
 case ${E2E_PARALLEL} in
-    'y'|'Y')           ginkgo_args+=("--nodes=25") ;;
-    [1-9]|[1-9][0-9]*) ginkgo_args+=("--nodes=${E2E_PARALLEL}") ;;
+    'y'|'Y'|'true')
+        # The flag '--p' will automatically detect the optimal number of ginkgo nodes.
+        ginkgo_args+=("--p")
+        # Skip serial tests if parallel mode is enabled.
+        E2E_SKIP="\\[Serial\\]|${E2E_SKIP}" ;;
 esac
 
-echo "/usr/local/bin/ginkgo ${ginkgo_args[@]} /usr/local/bin/e2e.test -- --disable-log-dump --repo-root=/kubernetes --provider=\"${E2E_PROVIDER}\" --report-dir=\"${RESULTS_DIR}\" --kubeconfig=\"${KUBECONFIG}\""
-/usr/local/bin/ginkgo "${ginkgo_args[@]}" /usr/local/bin/e2e.test -- --disable-log-dump --repo-root=/kubernetes --provider="${E2E_PROVIDER}" --report-dir="${RESULTS_DIR}" --kubeconfig="${KUBECONFIG}" | tee ${RESULTS_DIR}/e2e.log &
+ginkgo_args+=(
+    "--focus=${E2E_FOCUS}"
+    "--skip=${E2E_SKIP}"
+    "--noColor=true"
+)
+
+# Get the number of nodes to test in the cluster.
+# Any taint with effect:NoSchedule means a node requires special tolerations from deployments.
+NUM_NODES=$(kubectl get nodes -o=jsonpath='{range .items[*]}{.spec.taints}{"\n"}{end}' | grep -cv "effect:NoSchedule")
```

Comment:
So I like everything here, but let's take out this part and put it in the framework.